### PR TITLE
added serialization for classType, bigDecimalType, byteBuffer, metric…

### PR DESCRIPTION
**Pull request by vkagamlyk**
*Issue #, if available:*
https://bitquill.atlassian.net/browse/AN-1139

*Description of changes:*
[added serialization for classType, bigDecimalType, byteBuffer, metricsType, traversalMetricsType

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: gremlin-go/driver/graphBinary.go [hunk 1]
```diff
@@ -131,6 +135,20 @@ func listWriter(value interface{}, buffer *bytes.Buffer, typeSerializer *graphBi
...
+	v := value.(*ByteBuffer)
+	err := binary.Write(buffer, binary.BigEndian, int32(len(v.Data)))
+	if err != nil {
+		return nil, err
+	}
+	buffer.Write(v.Data)
+
+	return buffer.Bytes(), nil
+
+	return listWriter(v, buffer, typeSerializer)
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: Something weird here.

File: gremlin-go/driver/graphBinary.go [hunk 2]
```diff
@@ -290,38 +308,10 @@ func getSignedBytesFromBigInt(n *big.Int) []byte {
...
-	newBigInt.Neg(newBigInt)
-	return newBigInt
-}
-
 // Format: {length}{value_0}...{value_n}
 func bigIntWriter(value interface{}, buffer *bytes.Buffer, _ *graphBinaryTypeSerializer) ([]byte, error) {
-	v := value.(*big.Int)
-	signedBytes := getSignedBytesFromBigInt(v)
+	v := value.(big.Int)
+	signedBytes := getSignedBytesFromBigInt(&v)
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: Should we use reflection to check if ptr type and fall back if not?

File: gremlin-go/driver/graphBinary.go [hunk 3]
```diff
@@ -335,6 +325,22 @@ func bigIntWriter(value interface{}, buffer *bytes.Buffer, _ *graphBinaryTypeSer
...
+// Format: {scale}{unscaled_value}
+func bigDecimalWriter(value interface{}, buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer) ([]byte, error) {
+	v := value.(*BigDecimal)
+	err := binary.Write(buffer, binary.BigEndian, v.Scale)
+	if err != nil {
+		return nil, err
+	}
+
+	return bigIntWriter(v.UnscaledValue, buffer, typeSerializer)
+}
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: Having mismatch of pointers and not pointers for type assertion on the functions is odd

File: gremlin-go/driver/graphBinary.go [hunk 4]
```diff
@@ -675,6 +681,16 @@ func (serializer *graphBinaryTypeSerializer) getType(val interface{}) (dataType,
...
+	case *BigDecimal:
+		return bigDecimalType, nil
+	case *GremlinType:
+		return classType, nil
+	case Metrics:
+		return metricsType, nil
+	case TraversalMetrics:
+		return traversalMetricsType, nil
+	case *ByteBuffer:
+		return byteBuffer, nil
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: Same comment - mixture of pointers and non-pointers. I think the functions should take *type, type and then type assert appropriately from there.

File: gremlin-go/driver/traversal.go [hunk 5]
```diff
@@ -477,3 +479,32 @@ var WithOptions withOptions = withOptions{
...
+type GremlinType struct {
+	Fqcn string
+}
+
+type BigDecimal struct {
+	Scale         int32
+	UnscaledValue big.Int
+}
+
+type ByteBuffer struct {
```
Comment (lyndonbauto) [inline gremlin-go/driver/traversal.go]: Add comments in front of public structs for documentation. Also struct above this - WithOptions could use a comment as well (my bad missing that one).

File: gremlin-go/driver/graphBinary.go (hunk 1 above)
Comment (vkagamlyk) [reply gremlin-go/driver/graphBinary.go]: fixed in https://github.com/lyndonbauto/tinkerpop/pull/112/commits/86857c054213c55ffb3c0bc79468c8d158a4e1da

File: gremlin-go/driver/graphBinary.go (hunk 2 above)
Comment (vkagamlyk) [reply gremlin-go/driver/graphBinary.go]: added in https://github.com/lyndonbauto/tinkerpop/pull/112/commits/86857c054213c55ffb3c0bc79468c8d158a4e1da

File: gremlin-go/driver/graphBinary.go (hunk 3 above)
Comment (vkagamlyk) [reply gremlin-go/driver/graphBinary.go]: added in https://github.com/lyndonbauto/tinkerpop/pull/112/commits/86857c054213c55ffb3c0bc79468c8d158a4e1da

File: gremlin-go/driver/traversal.go (hunk 5 above)
Comment (vkagamlyk) [reply gremlin-go/driver/traversal.go]: added in https://github.com/lyndonbauto/tinkerpop/pull/112/commits/86857c054213c55ffb3c0bc79468c8d158a4e1da

File: gremlin-go/driver/graphBinary.go (hunk 4 above)
Comment (vkagamlyk) [reply gremlin-go/driver/graphBinary.go]: added in https://github.com/lyndonbauto/tinkerpop/pull/112/commits/86857c054213c55ffb3c0bc79468c8d158a4e1da

**Review by lyndonbauto — APPROVED** [review]
lgtm

**Comment by vkagamlyk** [comment]
moving to apache/tinkerpop

**Pull request closed without merging**